### ISSUE TITLE
Trigger based alert doesn't fire

### ISSUE THREAD

**Issue by DonLuigi** [redstone99/hass-alert2#7]
I recently discovered alert2 and it's so much better than built-in alert.

I do have an issue with trigger based alerts.
1. Condition is mandatory - why is that? Maybe I am missing something? In automations it is optional...
2. Following alert doesn't fire regardless of  sensor.entry_door_battery state:
```
  - domain: entry_door
    name: battery_full
    trigger:
      - platform: numeric_state
        entity_id: sensor.entry_door_battery
        above: 79
    condition: "{{ true }}
```
I could put above 79 in condition, but I don't want alert fired each time sensor.entry_door_battery increments by 1.

**Comment by redstone99** [comment]
Hi, thanks for giving Alert2 a spin. You're totally right. I just released v1.7.3 that fixes a few bugs:
1. `condition` is now optional when using triggers.
1. Triggers are now (hopefully) initialized correctly and so should fire as expected.  The code wasn't initializing triggers correctly and so an entity_id listed in a trigger, like "sensor.entry_door_battery", was being seen as a list of entities: ["s", "e", "n", "s", ...], none of which were found and so the trigger never triggered.   For some reason this wasn't affecting `platform: state` which is what I was testing with and so I missed the bug. 

Let me know if the fix works for you.  Feedback from real users is super valuable so let me know if you see anything else not working as expected!
-J

**Comment by DonLuigi** [comment]
It works. Thank you again for an excellent addon - this is what original alert should have been in the first place. I got rid of so many helpers, template sensors, automations and other rubbish, related to alerts, and pack all of the logic into alert2.
The best refactoring is when it results in significant reduction while maintaining functionality. Perfect.

I even managed to implement an alert that triggers upon condition, and deactivates manually. For condition I set something like
`condition: "{{ states ('binary_sensor.smoke_detectors') == 'on' or states('alert2.smoke_detectors_smoke_detected') == 'on' }}"`
pointing condition to state of the alert itself.

The alarm triggers when states ('binary_sensor.smoke_detectors') is 'on', can be acknowledged as needed, and finally turned off by running python script set_state via a tap_action, after checking all the cameras, etc.

This is quite a common pattern I use - alerts started via condition and dismissed manually. I wish I could directly dismiss (turn off) an alert via service call, but this workaround using python script set_state is also fine.

I use alert state many times as a state bearing entity, checking weather it is on or off, to hide / display stuff in cards and control automations.

BTW, I also found customer card to be helpful, but mainly for debugging and testing purposes - error checking and such. I use auto entities to collect active alerts and display them in a main card. Mainly because it gives me a cleaner look, and displays nothing if no alerts are active.

Thanks again for great work.

**Comment by redstone99** [comment]
Great, glad it works.  

Manual deactivation is an interesting use case. It's a form of hysteresis where once the alert has turned on, you want it to stay on even if the underlying firing condition is no longer true.  Your approach is clever and probably will continue working, though it does depend on some Alert2 implementation details that may change in the future.

Ideally, it'd probably be better to support your use case explicitly, say by adding a new config parameter, `manual_off`, indicating an alert will be turned off manually, and adding a new service call, `alert2.manual_turn_off` to turn a firing alert off.

**Comment by DonLuigi** [comment]
I don't want to impose, but it would absolutely be better to natively support manual deactivation. Not only because it would be a cleaner use of the integration (no need for python hack), but also because implementation would spill over to trigger based alerts as well.

I'll give you an example:
### Alert if I forgot to charge my ebike after arriving home
This is implemented by having a one hour timer that starts when I arrive home. The automation cancels the timer if power consumption in smart socket rises to about 100W. But if the timer manages to finish, I use that as a trigger for an alert. Which is great and all, but that alert is not active after it fires, even though I didn't fix the problem. Ideally, it should stay active until I manage to get my butt down to the garage and attach the battery to the charger. Only then it would be appropriate to deactivate the alert.

What I am trying to say is, having a manual deactivation option makes sense for both, condition and trigger based alerts. It is useful when there are conditions and actions required to deactivate an alert that cannot be modeled in HA. Like a person checking cameras, or attaching a charger to a battery.

On another note, there is one small improvement I would propose: different icons for acknowledged and unacknowledged alerts. I am using auto entities card instead of your custom card in my main dashboard, and it's simply not possible to differentiate between the both types.

Finally I have to say I am really fortunate I discovered your integration. Seems to me HA has been neglecting this area for so long. A clear indication of this is the fact that you have to fully restart HA (not just YAML config) when making changes to builtin alert integration. But it's a really important area. I mean, it's of no use having all those fancy automations, but not realizing sensors dying, glitching, dropping out of network, batteries going empty, etc. I was especially impressed the way you designed generator patterns. Currently I have 30 alerts. I can't imagine configuring them one by one.

Happy holidays and wish you all the best

**Comment by mitman98** [comment]
Hi @DonLuigi, completely agreed on this integration and it's importance - having so much tied into HA and then not having the ability to know that something is awry in the system is a big gap in the core code. Alert2 has been a game changer for me. Thanks for your thoughts on it, Josh has been great at really taking user feedback into account.

Your example is a great example. I also like the idea that if trigger x happens, let's ensure that someone checks the camera/etc. in a certain amount of time, which is an event followed by another event.

I don't disagree that there's some validity in having the ability to allow trigger alerts to be in an "active" state after triggering and manually deactivate them.

A few thoughts come to mind:

1. I think there is a general use case of having an alert that is fired via a trigger or service call (as implemented today) and then deactivated by a separate trigger or service call. I'd suggest both ways of deactivation are important.
1. If the deactivation is never manually fired (due to an automation error for example), I think there likely would need to be some override control to also user manually deactivate at the Alert2 level - I think in the Alert2 UI info popup may make sense, as otherwise we'd be forcing users to call service calls to deactivate trigger alerts which I don't think is optimal. There's then some risk a user manually deactivates an alert through the UI that shouldn't have been but I think the tradeoff is correct.
2. However, I'm not sure deactivation of a conditional alert makes semantic sense. I think ack/snooze/disable are more of the mechanisms that may make sense, because the underlying condition is still true - we just don't want to get alerted any more, with one caveat, below:
a. Usually, if the condition is still active I want to know in the UI. I can ack/snooze/disable if I don't want to get further notifications right now and depending on the mission criticalness of the condition being alerted on - although snoozing and disabling can be somewhat brute force as for snooze I have to pick a time period for which I don't care about and disable has me checking the UI from time to time (which is fine for conditions that are "nice to fix" and not reason for alarm). If you have a manual deactivation trigger/service call of a conditional alert, you would end up in a novel state where the alert is deactivated by the manual step but the condition is still true. You'd want a special indication of this state on the alert object and in the UI because presumably with the feature you're not going to re-fire the alert until the condition goes false and then true again so there's a unique state the alert is in for this time period. 
b. Ack/snooze/disable kind of handles this state already; however, it's not tied to the condition itself, only the the user of the UI. So if the condition resolves and then occurs again you can't link that up to a new activation of the alert.
c. I think there is a gap of a generalized requirement of I want to "mark this alert as over and start notifying agian when the condition becomes true again". I'm curious if this is better represented in the system and to the user of the UI as (i) at the alert level: the alert is "over" and it is now waiting for the condition to go false before the alert is primed again, or (ii) at the notification level on top of the alert: the alert itself is still active but notifications are "snoozed until active" for example. It's a slight semantic difference, curious your thoughts.

Perhaps we enumerate some potential use cases for conditional alerts that can be deactivated and also what that state is of alert has been deactivated but condition is still true.

**Comment by flimofly** [comment]
> Hi, thanks for giving Alert2 a spin. You're totally right. I just released v1.7.3 that fixes a few bugs:
> 
> 1. `condition` is now optional when using triggers.
> 2. Triggers are now (hopefully) initialized correctly and so should fire as expected.  The code wasn't initializing triggers correctly and so an entity_id listed in a trigger, like "sensor.entry_door_battery", was being seen as a list of entities: ["s", "e", "n", "s", ...], none of which were found and so the trigger never triggered.   For some reason this wasn't affecting `platform: state` which is what I was testing with and so I missed the bug.
> 
> Let me know if the fix works for you. Feedback from real users is super valuable so let me know if you see anything else not working as expected! -J

I just upgraded to v1.7.3 but the below doesn't work still. Or is there a mistake in the syntax that I'm not seeing?

```
alert2:
  alerts:
    - domain: washing machine
      name: wet_laundry
      trigger:
        - platform: state
          entity_id: sensor.washingmachine_ready
      condition: "{{ is_state('binary_sensor.doorsensor_opening','closed') }}"
      message: "Laundry done - wet clothes in machine!"
      notifier: mobile_app_iphone
```

**Comment by DonLuigi** [comment]
Why not try something like this, without condition:
```
  - platform: state
    entity_id: binary_sensor.doorsensor_opening
    to: closed
```
BTW, I don't understand relationship between sensor.washingmachine_ready and binary_sensor.doorsensor_opening? Are they both contributing to when alert should be activated? In which state? You defined that for binary_sensor.doorsensor_opening, but not for sensor.washingmachine_ready.

Anyway, the easiest way to test is to make an automation, and if trigger works there, you can copy paste it into alert config.

**Comment by flimofly** [comment]
I try to make use of alert2's functionality of firing as long as particular condition is met.  I want the alert to fire every few minutes when the washing machine is ready (sensor.washingmachine_ready = on) until the door is opened (binary_sensor.doorsensor_opening = closed). 

For testing purposes I created a switch so I can turn sensor.washingmachine_ready on and off.  The doorsensor status is closed. Even then, the alert does not fire.

**Comment by DonLuigi** [comment]
I don't think that will work. State trigger fires when state changes. In your case, none of the states are changing between 'on' and 'opened'.

In other words, if you put 'on' state in your trigger (it's missing in your original example) you should be able to get alert once (when washing machine becomes ready), but that is about it.

What would work tough is a suggestion from @mitman98 where alert would get a new optional deactivation trigger. This, or having a manual deactivation, effectively changes an alert duration from an instant (as is today) to a duration.

Now that's when alert2 mechanisms of recurring alerts (reminder_frequency_mins, ...) would (should) kick in.

**Comment by mitman98** [comment]
This is an interesting use case. You want the alert to begin functioning when the washing machine is done, and continue firing when another condition is still true, regardless of the washing machine done state. So I do think it's a trigger based alert as you've identified because you only care about the initial state of the washing machine being done changing, then no longer care - you now just care about the ongoing condition of the door closed sensor. 

And you want to be regularly notified until the condition of the door being closed is over. As @DonLuigi said this is what reminder_frequency_mins does on a condition based alert - sends a new notification every x minutes until the condition is false.

However, I'm not fully certain but my understanding is that conditions are only checked for trigger/event-based alerts at the time of  the event and are not an ongoing check like with condition only alerts, so your use case doesn't sound like it'll work under the current design (@redstone99 might also be able to confirm).

@redstone99 I think there's a valid use case extension of event based alerts - actually making them a condition alert that can only be activated if the trigger happens but then stay active until the condition stops.

Again, this use case is great timing as we're thinking about the next set of features because there's appearing to be parallels of how alerts can be activated and deactivated (combinations of events and/or condition states).

**Comment by flimofly** [comment]
Helpful explanation! I thought that perhaps a way out would be to switch the condition and the trigger around. In other words: when the door is closed (= state change), check whether the washing machine status is set to finished.  I understand, however, that the condition is checked only once, i.e. when the state changes, so this wouldn't be a solution (apart from not being very intuitive).

**Comment by DonLuigi** [comment]
Example with the washing machine is a really good one, that cannot be implemented using condition alerts alone. You could of course define condition (sensor.washingmachine_ready = on and binary_sensor.doorsensor_opening = closed), but this would start alerting again as soon as you would close the door after emptying it. Unless you have additional automation that will clear sensor.washingmachine_ready when binary_sensor.doorsensor_opening = opened.
Not ideal.

With having ability to configure both triggers:
1. alert activation trigger: when power consumption in smart socket falls to below 5W
2. alert deactivation trigger: binary_sensor.doorsensor_opening changes state to open
and reminder_frequency_mins: 1
you need no helpers, automations, etc,

Bottom line: the only way this alert is doable is if we have activate and deactivate triggers.

My example alert when I forget to charge my ebike is very similar. It activates when timer finishes. But it deactivates when power reaches around 100W. And in the meantime: reminder_frequency_mins: 1

**Comment by mitman98** [comment]
I think the issue is more fundamental - Alert2 trigger alerts are not stateful. They fire once at the time the event occurs. They do not track an on/off state (their state is the last fired time I believe), and that's it whereas condition alerts are stateful. To implement ongoing checks would require a change to the architecture of the event based alerts at which time you'd just want to track the condition as an ongoing state.

If it helps I think this YAML with a binary_sensor helper and two automations to turn it on/off might do the trick for you to have a binary_sensor that tracks the period of concern for your use case: washing machine needs to be emptied (washing machine is done, door is still closed):

input_boolean.washing_machine_needs_emptying should be true during the time period when your washing machine has finished and you haven't opened the door. 

You can then create a condition alert in Alert2 that just alerts on input_boolean.washing_machine_needs_emptying being on, and set a reminder frequency to notify you periodically.

```
input_boolean:
  washing_machine_needs_emptying:
    name: Washing Machine Needs Emptying
    initial: false

alias: Washing Machine Done
trigger:
  - platform: state
    entity_id: sensor.washing_machine_ready 
    to: 'on'
condition:
  - condition: state
    entity_id: binary_sensor.door_sensor
    state: 'off'  # Door is closed when the washing machine is ready (his condition is not fully required)
action:
  - service: input_boolean.turn_on
    target:
      entity_id: input_boolean.washing_machine_needs_emptying

alias: Washing Machine Emptied Rest
trigger:
  - platform: state
    entity_id: binary_sensor.door_sensor
    to: 'on'  # Door opens
condition:
  - condition: state
    entity_id: input_boolean.washing_machine_needs_emptying
    state: 'on'  # Only act if the washing machine was done
action:
  - service: input_boolean.turn_off
    target:
      entity_id: input_boolean.washing_machine_needs_emptying
```

**Comment by redstone99** [comment]
Hi All, interesting conversation and use cases! Here's my take:

My mental model for alerting is that an alert should be firing while a problem exists, and Alert2 should support being able to capture the time interval of the problem.  So far, we have event alerts, which indicate a momentary problem and condition alerts which indicate a problem that exists while a condition is true.

Perhaps we need a new type of alert, which I'll call a "latching" alert. The main feature is that the triggers or conditions for turning the alert on and off may differ.  So when it starts firing, it "latches" the firing state and keeps on firing until separate trigger/conditions are met for it to stop firing. Basically what @DonLuigi described.

To express the use cases raised here in terms of a "latching" alert:
* An alert triggers on when `binary_sensor.smoke_detectors` becomes true and turns off manually (by a service call or button in the more-info UI)
* An alert triggers on when `sensor.washingmachine_ready` becomes true and turns off when `binary_sensor.doorsensor_opening` becomes true
* An alert triggers on when sensor.away_status == "home" with `delay_on_secs`=1 hour.  It triggers off when `sensor.smart_socket_watts` > 100.  I'm postulating that there are reasonable semantics for how `delay_on_secs` would work with "latching" alerts.

At least in my mind, I try to keep distinct the concept of when a problem exists and an alert should be firing, and how notifications or UI display are managed.  I think notifications&UI are more the domain of @mitman98 's "staged alerts" suggestion.  (I'm hoping to post more thoughts on staged alerting in the HA forum thread).

Thoughts?
-Josh

**Comment by DonLuigi** [comment]
To me at least, conceptually all those cases are basically the same. If I were writing alert2, that's how I would set it up anyway:
Every alert has only two key parameters:
1. event to activate and
2. event to deactivate

They are defined in every single alert.

One way of describing both events is using a single condition - as long as it is true, alert is active. So a single condition describes event to activate (condition switches from false to true) and to deactivate (condition switches from true to false)

The other is providing explicit triggers.

Providing only activation trigger is just a shorthand notification for providing the same trigger for activation and deactivation. Meaning, alert deactivates as soon as it activates, if trigger is the same. Conceptually, event notification alerts are just a subset of all alerts.

Having a manual deactivation again is just a shorthand for having an (not ideal) external helper and triggering off it. But hey, we already have manual activation - it think it's called report. Same thing, just different direction. It's just about defining activate and deactivate events.

If both are provided - condition and trigger(s) - then condition switches from defining activation and deactivation events to filtering duty for triggers that actually define activation and deactivation events.

In the end, there are no different modes of alerts and code is the same for all cases. As such it is easier to code, describe, and is the most flexible to use. I think this would be pretty backward compatible setup.

As far as staged alerts go, if I understand correctly, they are just fine graining of active state of an alert. So instead of having a single 'on' state for the whole duration, there is some kind of simplified state machine with transitions between flavors of 'on' state. As such it has nothing to do with above postulates. It is an independent feature that's not really affected by what's discussed above.

**Comment by mitman98** [comment]
I generally agree with your abstraction @DonLuigi. I do suggest moving to this thread in the HA Community as there are several GitHub issues related to the work being done and the overall conversation is also progressing there... https://community.home-assistant.io/t/alert2-a-new-alerting-component/654597/115?u=tman98

**Comment by redstone99** [comment]
@DonLuigi  - good summary. I'm deep in the UI alert editing work at the moment, but was thinking along your lines for implementation as well.

**Comment by redstone99** [comment]
I just released support for split on/off conditions & triggers, including manual triggering. @DonLuigi , can you give it a whirl and see if it meets your needs?  If so I can close this.
J

**Comment by DonLuigi** [comment]
Excellent. I really mean it. This new version is powerful and versatile.
And backward compatible. None of my older alerts stopped functioning after installing 1.9.
Thank you for the effort and for listening to your user base.
Alert2 must be one of the finest integrations for HA. Really big thumbs up ;-)

**Comment by redstone99** [comment]
Glad you're getting a lot of value of it, and thanks for your input to help make it better!
